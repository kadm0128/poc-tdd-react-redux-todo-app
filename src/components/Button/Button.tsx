import { ComponentPropsWithoutRef, ReactNode } from "react";
import { getClassNames } from "../../utils";
import "./styles.css";
import { Color, Size, Variant } from "./Button.types";

type Props = ComponentPropsWithoutRef<"button"> & {
   color?: Color;
   variant?: Variant;
   size?: Size;
   className?: string;
   children: ReactNode;
};

const Button = (props: Props) => {
   const { color = 'primary', variant = 'contained', size = 'medium', className = '', children } = props;
   const classNames = getClassNames('button', [`${variant}-${color}`, size], className);

   return (
      <button className={classNames}>
         {children}
      </button>
   );
}

export default Button