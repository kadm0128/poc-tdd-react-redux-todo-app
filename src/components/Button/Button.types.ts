export type Color = 'primary' | 'secondary' | 'success' | 'info' | 'warning' | 'error';
export type Size = 'small' | 'medium' | 'large';
export type Variant = 'contained' | 'outlined'| 'text';