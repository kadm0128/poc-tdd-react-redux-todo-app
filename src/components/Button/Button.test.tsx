import { render, screen, renderer } from "../../utils//test-utils";
import Button from './Button';
import { Color, Size, Variant } from "./Button.types";

describe('BUTTON', () => {
   describe('Component Rendering and Props Handling', () => {
      test('renders the button text correctly', () => {
         render(<Button>Click me</Button>);
         const button = screen.getByText('Click me');
         expect(button).toBeInTheDocument();
      });

      test('applies default styles and classes correctly', () => {
         const { container } = render(<Button>Click me</Button>);
         const button = container.querySelector('button');
         expect(button).toHaveClass('dc-button dc-button-contained-primary dc-button-medium');
      });

      describe('renders the specified color correctly', () => {
         const supportedColors: Color[] = ['primary', 'secondary', 'success', 'info', 'warning', 'error'];

         test.each(supportedColors)('renders with specified color - %s', (color: Color) => {
            const { container } = render(<Button color={color}>Click me</Button>);
            const button = container.querySelector('button');
            expect(button).toHaveClass(`dc-button dc-button-contained-${color} dc-button-medium`);
         });
      });

      describe('renders the specified variant correctly', () => {
         const supportedVariants: Variant[] = ['contained', 'outlined', 'text'];

         test.each(supportedVariants)('renders with specified variant - %s', (variant: Variant) => {
            const { container } = render(<Button variant={variant}>Click me</Button>);
            const button = container.querySelector('button');
            expect(button).toHaveClass(`dc-button dc-button-${variant}-primary dc-button-medium`);
         });
      });

      describe('renders the specified size correctly', () => {
         const supportedSizes: Size[] = ['small', 'medium', 'large'];

         test.each(supportedSizes)('renders with specified size - %s', (size: Size) => {
            const { container } = render(<Button size={size}>Click me</Button>);
            const button = container.querySelector('button');
            expect(button).toHaveClass(`dc-button dc-button-contained-primary dc-button-${size}`);
         });
      });

      test('renders the given set of props correctly', () => {
         const { container } = render(<Button size="small" color="success" variant="outlined">Click me</Button>);
         const button = container.querySelector('button');
         expect(button).toHaveClass('dc-button dc-button-outlined-success dc-button-small');
      });
   });

   // Test interactions within component (testing event handlers, state changes, and component updates triggered by user interactions)
   describe('Component Interactions and State Management', () => {
      test('test', () => {

      });

      test('test', () => {

      })
   });

   describe('Component Error Handling', () => {
      test('test', () => {

      })
   });

   describe('Snapshot Testing', () => {
      const supportedSizes: Size[] = ['small', 'medium', 'large'];
      const supportedVariants: Variant[] = ['contained', 'outlined', 'text'];
      const supportedColors: Color[] = ['primary', 'secondary', 'success', 'info', 'warning', 'error'];

      for (const color of supportedColors) {
         for (const variant of supportedVariants) {
            for (const size of supportedSizes) {
               test(`Snapshot Testing - ${color}, ${variant}, ${size}`, () => {
                  const tree = renderer.create(<Button size={size} color={color} variant={variant}>Click me</Button>).toJSON();
                  expect(tree).toMatchSnapshot();
               });
            }
         }
      }
   });

   describe('Integration Testing', () => {
      test('test', () => {

      })
   });

   describe('Accessibility', () => {
      test('test', () => {

      })
   });

   describe('Edge Cases', () => {
      test('test', () => {

      })
   });
})