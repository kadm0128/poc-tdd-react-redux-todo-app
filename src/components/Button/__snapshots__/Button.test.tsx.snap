// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, contained, large 1`] = `
<button
  className="dc-button dc-button-contained-error dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, contained, medium 1`] = `
<button
  className="dc-button dc-button-contained-error dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, contained, small 1`] = `
<button
  className="dc-button dc-button-contained-error dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, contained, smalll 1`] = `
<button
  className="dc-button dc-button-contained-error dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, outlined, large 1`] = `
<button
  className="dc-button dc-button-outlined-error dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, outlined, medium 1`] = `
<button
  className="dc-button dc-button-outlined-error dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, outlined, small 1`] = `
<button
  className="dc-button dc-button-outlined-error dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, outlined, smalll 1`] = `
<button
  className="dc-button dc-button-outlined-error dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, text, large 1`] = `
<button
  className="dc-button dc-button-text-error dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, text, medium 1`] = `
<button
  className="dc-button dc-button-text-error dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, text, small 1`] = `
<button
  className="dc-button dc-button-text-error dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - error, text, smalll 1`] = `
<button
  className="dc-button dc-button-text-error dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, contained, large 1`] = `
<button
  className="dc-button dc-button-contained-info dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, contained, medium 1`] = `
<button
  className="dc-button dc-button-contained-info dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, contained, small 1`] = `
<button
  className="dc-button dc-button-contained-info dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, contained, smalll 1`] = `
<button
  className="dc-button dc-button-contained-info dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, outlined, large 1`] = `
<button
  className="dc-button dc-button-outlined-info dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, outlined, medium 1`] = `
<button
  className="dc-button dc-button-outlined-info dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, outlined, small 1`] = `
<button
  className="dc-button dc-button-outlined-info dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, outlined, smalll 1`] = `
<button
  className="dc-button dc-button-outlined-info dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, text, large 1`] = `
<button
  className="dc-button dc-button-text-info dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, text, medium 1`] = `
<button
  className="dc-button dc-button-text-info dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, text, small 1`] = `
<button
  className="dc-button dc-button-text-info dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - info, text, smalll 1`] = `
<button
  className="dc-button dc-button-text-info dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, contained, large 1`] = `
<button
  className="dc-button dc-button-contained-primary dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, contained, medium 1`] = `
<button
  className="dc-button dc-button-contained-primary dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, contained, small 1`] = `
<button
  className="dc-button dc-button-contained-primary dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, contained, smalll 1`] = `
<button
  className="dc-button dc-button-contained-primary dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, outlined, large 1`] = `
<button
  className="dc-button dc-button-outlined-primary dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, outlined, medium 1`] = `
<button
  className="dc-button dc-button-outlined-primary dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, outlined, small 1`] = `
<button
  className="dc-button dc-button-outlined-primary dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, outlined, smalll 1`] = `
<button
  className="dc-button dc-button-outlined-primary dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, text, large 1`] = `
<button
  className="dc-button dc-button-text-primary dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, text, medium 1`] = `
<button
  className="dc-button dc-button-text-primary dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, text, small 1`] = `
<button
  className="dc-button dc-button-text-primary dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - primary, text, smalll 1`] = `
<button
  className="dc-button dc-button-text-primary dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, contained, large 1`] = `
<button
  className="dc-button dc-button-contained-secondary dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, contained, medium 1`] = `
<button
  className="dc-button dc-button-contained-secondary dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, contained, small 1`] = `
<button
  className="dc-button dc-button-contained-secondary dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, contained, smalll 1`] = `
<button
  className="dc-button dc-button-contained-secondary dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, outlined, large 1`] = `
<button
  className="dc-button dc-button-outlined-secondary dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, outlined, medium 1`] = `
<button
  className="dc-button dc-button-outlined-secondary dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, outlined, small 1`] = `
<button
  className="dc-button dc-button-outlined-secondary dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, outlined, smalll 1`] = `
<button
  className="dc-button dc-button-outlined-secondary dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, text, large 1`] = `
<button
  className="dc-button dc-button-text-secondary dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, text, medium 1`] = `
<button
  className="dc-button dc-button-text-secondary dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, text, small 1`] = `
<button
  className="dc-button dc-button-text-secondary dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - secondary, text, smalll 1`] = `
<button
  className="dc-button dc-button-text-secondary dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, contained, large 1`] = `
<button
  className="dc-button dc-button-contained-success dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, contained, medium 1`] = `
<button
  className="dc-button dc-button-contained-success dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, contained, small 1`] = `
<button
  className="dc-button dc-button-contained-success dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, contained, smalll 1`] = `
<button
  className="dc-button dc-button-contained-success dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, outlined, large 1`] = `
<button
  className="dc-button dc-button-outlined-success dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, outlined, medium 1`] = `
<button
  className="dc-button dc-button-outlined-success dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, outlined, small 1`] = `
<button
  className="dc-button dc-button-outlined-success dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, outlined, smalll 1`] = `
<button
  className="dc-button dc-button-outlined-success dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, text, large 1`] = `
<button
  className="dc-button dc-button-text-success dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, text, medium 1`] = `
<button
  className="dc-button dc-button-text-success dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, text, small 1`] = `
<button
  className="dc-button dc-button-text-success dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - success, text, smalll 1`] = `
<button
  className="dc-button dc-button-text-success dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, contained, large 1`] = `
<button
  className="dc-button dc-button-contained-warning dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, contained, medium 1`] = `
<button
  className="dc-button dc-button-contained-warning dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, contained, small 1`] = `
<button
  className="dc-button dc-button-contained-warning dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, contained, smalll 1`] = `
<button
  className="dc-button dc-button-contained-warning dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, outlined, large 1`] = `
<button
  className="dc-button dc-button-outlined-warning dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, outlined, medium 1`] = `
<button
  className="dc-button dc-button-outlined-warning dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, outlined, small 1`] = `
<button
  className="dc-button dc-button-outlined-warning dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, outlined, smalll 1`] = `
<button
  className="dc-button dc-button-outlined-warning dc-button-smalll"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, text, large 1`] = `
<button
  className="dc-button dc-button-text-warning dc-button-large"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, text, medium 1`] = `
<button
  className="dc-button dc-button-text-warning dc-button-medium"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, text, small 1`] = `
<button
  className="dc-button dc-button-text-warning dc-button-small"
>
  Click me
</button>
`;

exports[`BUTTON > Snapshot Testing > Snapshot Testing - warning, text, smalll 1`] = `
<button
  className="dc-button dc-button-text-warning dc-button-smalll"
>
  Click me
</button>
`;
